third_party/move/move-vm/runtime/src/loader/mod.rs

```diff
@@ -793,12 +824,134 @@ impl Loader {
         })
     }
 
+    /// Traverses the whole transitive closure of dependencies, starting from the specified
+    /// modules and perform gas metering.
+    ///
+    /// The traversal follows a depth-first order, with the module itself being visited first,
+    /// followed by its dependencies, and finally its friends.
+    /// Changing this order without a valid reason may affect the gas semantics and must be
+    /// avoided.
+    ///
+    /// This will result in the shallow-loading of the modules -- they will be read from the
+    /// storage as bytes and then deserialized, but NOT converted into the runtime representation.
+    ///
+    /// It should also be noted that this is implemented in a way that avoids the cloning of
+    /// `ModuleId`, a.k.a. heap allocations, as much as possible, which is critical for
+    /// performance.
+    ///
+    /// TODO: Revisit the order of traversal. Consider switching to alphabetical order.
+    pub(crate) fn check_dependencies_and_charge_gas<'a, I>(
+        &self,
+        module_store: &ModuleStorageAdapter,
+        data_store: &mut TransactionDataCache,
+        gas_meter: &mut impl GasMeter,
+        visited: &mut BTreeMap<(&'a AccountAddress, &'a IdentStr), ()>,
+        referenced_modules: &'a Arena<Arc<CompiledModule>>,
+        ids: I,
+    ) -> VMResult<()>
+    where
+        I: IntoIterator<Item = (&'a AccountAddress, &'a IdentStr)>,
+        I::IntoIter: DoubleEndedIterator,
+    {
+        // Initialize the work list (stack) and the map of visited modules.
+        let mut stack = Vec::with_capacity(512);
```

Comment:
this has to be larger than 420, right? is there a way to access that parameter here?

Reply:
> is there a way to access that parameter here?

No easy way. The Move VM does not have access to the Aptos gas schedule. I'd say for now it's fine to use a hardcoded value here. The worst that can happen is a reallocation, in case we decide to increase the limit above 512.

Reply:
it's only for perf, but still, if managing stack was important for not allocating things then getting while vector resizing that would be hard to avoid in the future because of a magic constant here would be not ideal?
Can we pass it as a parameter to this method?

Reply:
> if managing stack was important for not allocating things then getting while vector resizing that would be hard to avoid in the future because of a magic constant here would be not ideal? 
Can we pass it as a parameter to this method?

We can add a parameter to the vm config, but this particular optimization is very insignificant compared to the savings from not having to allocate new ModuleIds, so I'd not prioritize perfecting it right now.

Reply:
but this is how we end up with magic constants in the code, which like a convention we shouldn't do? if it is not a performance issue right now, we can remove 512, if it can become a performance issue later then the problem is not even knowing about it. I don't want to be a stickler though, agree on priorities - extracting as constant isn't too useful, but let's add a // TODO: determine based on vm config - max number of dependencies.

Reply:
I would still extract the value into a constant with a self-documenting name:
`const INITIAL_STACK_SIZE = 512`;